plugins/filter_aws/aws.c

```diff
@@ -237,21 +275,69 @@ static int get_metadata(struct flb_filter_aws *ctx, char *metadata_path,
         return -1;
     }
 
-    tmp = flb_sds_create_len(client->resp.payload, client->resp.payload_size);
+    if (key != NULL) {
+        /* get the value of the key from payload json string */
+        tmp = flb_json_get_val(client->resp.payload, client->resp.payload_size, key);
+        if (!tmp) {
+            tmp = flb_sds_create_len("NULL", 4);
+            flb_plg_info(ctx->ins, "ec2-metadata: %s is undefined in EC2 instance", key);
+        }
+    } else {
+        tmp = flb_sds_create_len(client->resp.payload, client->resp.payload_size);
+    }
+
     if (!tmp) {
         flb_errno();
         flb_http_client_destroy(client);
         flb_upstream_conn_release(u_conn);
         return -1;
     }
+
     *metadata = tmp;
-    *metadata_len = client->resp.payload_size;
+    *metadata_len = key == NULL ? client->resp.payload_size : strlen(tmp);
 
     flb_http_client_destroy(client);
     flb_upstream_conn_release(u_conn);
     return 0;
 }
 
+static int get_metadata(struct flb_filter_aws *ctx, char *metadata_path, flb_sds_t *metadata, size_t *metadata_len)
+{
+    return get_metadata_by_key(ctx, metadata_path, metadata, metadata_len, NULL);
+}
+
+/* get VPC metadata, it called IMDS twice. First is for getting the Mac ID and combine into the path for VPC.
+   Second call is using the VPC path to get the VPC id
+*/
+static int get_vpc_metadata(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t mac_id = NULL;
+    size_t len = 0;
+
+    /* get EC2 instance Mac id first before getting VPC id */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_MAC_PATH, &mac_id, &len);
+
+    if (ret < 0) {
+        flb_sds_destroy(mac_id);
+        return -1;
+    }
+
+    /* the VPC full path should be like: /latest/meta-data/network/interfaces/macs/{mac_id}/vpc-id/" */
+    flb_sds_t vpc_path = flb_sds_create("/latest/meta-data/network/interfaces/macs/");
+    vpc_path = flb_sds_cat(vpc_path, mac_id, len);
+    vpc_path = flb_sds_cat(vpc_path, "/vpc-id/", strlen("/vpc-id/"));
```

Comment:
This is a job for flb_sds_printf.

Search through the existing code base and you will find examples of how to use it

Reply:
Oh good to know! It's more appropriate by using flb_sds_printf, changed to it.